### ISSUE TITLE
dotnet test host process crashed on Ubuntu 18.04

### ISSUE THREAD

**Issue by wli3** [dotnet/runtime#1605]
_From @rkeithhill on Friday, January 10, 2020 5:50:30 AM_

We have some xUnit tests that are running fine on Windows and RedHat7.  The same tests cause a cryptic failure on Ubuntu 18.04:
```
Using dotnet v3.1.100 at path /usr/bin/dotnet
Output folder set to: /home/mctcbld/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin
Build started 1/9/2020 8:05:38 PM.
Test run for /home/mctcbld/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/Foo.Api.UnitTest.dll(.NETCoreApp,Version=v3.1)
Microsoft (R) Test Execution Command Line Tool Version 16.3.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (64-bit .NET Core 3.1.0)
[xUnit.net 00:00:00.40]   Discovering: FooApi.UnitTest
[xUnit.net 00:00:00.46]   Discovered:  Foo.Api.UnitTest
[xUnit.net 00:00:00.46]   Starting:    Foo.Api.UnitTest
Checking for /tmp/FooApiUnitTestLicenses-es-foobuild-ub/005056901801/FOOF101-1FP.lic
Copying to /home/mctcbld/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/CSharpTestLicenses/FOOF101-1FP.lic
Set process env var FOOTEST_LICENSE_FILE to /home/mctcbld/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/CSharpTestLicenses from DesktopFooTests.cs
The active test run was aborted. Reason: Test host process crashed

Test Run Aborted.

Build FAILED.
    0 Warning(s)
    0 Error(s)
```
As you can see, there's really no error details here.  With `LD_DEBUG="all"`, I see this error show up:
```
      2367:     binding file /usr/share/dotnet/shared/Microsoft.NETCore.App/3.1.0/System.Native.so [0] to /lib/x86_64-linux-gnu/libc.so.6 [0]: normal symbol `ioctl' [GLIBC_2.2.5]
                                                                                                                        The active test run was aborted. Reason: Test host process crashed :       2383:        symbol=chdir;  lookup in file=dotnet [0]
      2383:     symbol=chdir;  lookup in file=/lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     symbol=chdir;  lookup in file=/lib/x86_64-linux-gnu/libdl.so.2 [0]
      2383:     symbol=chdir;  lookup in file=/usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0]
      2383:     symbol=chdir;  lookup in file=/lib/x86_64-linux-gnu/libm.so.6 [0]
      2383:     symbol=chdir;  lookup in file=/lib/x86_64-linux-gnu/libgcc_s.so.1 [0]
      2383:     symbol=chdir;  lookup in file=/lib/x86_64-linux-gnu/libc.so.6 [0]
      2383:     binding file /usr/share/dotnet/shared/Microsoft.NETCore.App/3.1.0/System.Native.so [0] to /lib/x86_64-linux-gnu/libc.so.6 [0]: normal symbol `chdir' [GLIBC_2.2.5]
      2383:     symbol=execve;  lookup in file=dotnet [0]
      2383:     symbol=execve;  lookup in file=/lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     symbol=execve;  lookup in file=/lib/x86_64-linux-gnu/libdl.so.2 [0]
      2383:     symbol=execve;  lookup in file=/usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0]
      2383:     symbol=execve;  lookup in file=/lib/x86_64-linux-gnu/libm.so.6 [0]
      2383:     symbol=execve;  lookup in file=/lib/x86_64-linux-gnu/libgcc_s.so.1 [0]
      2383:     symbol=execve;  lookup in file=/lib/x86_64-linux-gnu/libc.so.6 [0]
      2383:     binding file /usr/share/dotnet/shared/Microsoft.NETCore.App/3.1.0/System.Native.so [0] to /lib/x86_64-linux-gnu/libc.so.6 [0]: normal symbol `execve' [GLIBC_2.2.5]
      2383:
      2383:     file=libpthread.so.0 [0];  needed by /usr/share/dotnet/dotnet [0]
      2383:     find library=libpthread.so.0 [0]; searching
      2383:      search cache=/etc/ld.so.cache
      2383:       trying file=/lib/x86_64-linux-gnu/libpthread.so.0
      2383:
      2383:     file=libpthread.so.0 [0];  generating link map
      2383:       dynamic: 0x00007fa1eac1bd68  base: 0x00007fa1eaa02000   size: 0x000000000021e480
      2383:         entry: 0x00007fa1eaa08230  phdr: 0x00007fa1eaa02040  phnum:                  9
      2383:
      2383:
      2383:     file=libdl.so.2 [0];  needed by /usr/share/dotnet/dotnet [0]
      2383:     find library=libdl.so.2 [0]; searching
      2383:      search cache=/etc/ld.so.cache
      2383:       trying file=/lib/x86_64-linux-gnu/libdl.so.2
      2383:
      2383:     file=libdl.so.2 [0];  generating link map
      2383:       dynamic: 0x00007fa1eaa00d90  base: 0x00007fa1ea7fe000   size: 0x0000000000203110
      2383:         entry: 0x00007fa1ea7fee50  phdr: 0x00007fa1ea7fe040  phnum:                  7
      2383:
      2383:
      2383:     file=libstdc++.so.6 [0];  needed by /usr/share/dotnet/dotnet [0]
      2383:     find library=libstdc++.so.6 [0]; searching
      2383:      search cache=/etc/ld.so.cache
      2383:       trying file=/usr/lib/x86_64-linux-gnu/libstdc++.so.6
      2383:
      2383:     file=libstdc++.so.6 [0];  generating link map
      2383:       dynamic: 0x00007fa1ea7f6cc0  base: 0x00007fa1ea41e000   size: 0x00000000003df5c0
      2383:         entry: 0x00007fa1ea4b7180  phdr: 0x00007fa1ea41e040  phnum:                  8
      2383:
      2383:
      2383:     file=libm.so.6 [0];  needed by /usr/share/dotnet/dotnet [0]
      2383:     find library=libm.so.6 [0]; searching
      2383:      search cache=/etc/ld.so.cache
      2383:       trying file=/lib/x86_64-linux-gnu/libm.so.6
      2383:
      2383:     file=libm.so.6 [0];  generating link map
      2383:       dynamic: 0x00007fa1ea41cd90  base: 0x00007fa1ea080000   size: 0x000000000039d158
      2383:         entry: 0x00007fa1ea08ba80  phdr: 0x00007fa1ea080040  phnum:                  7
      2383:
      2383:
      2383:     file=libgcc_s.so.1 [0];  needed by /usr/share/dotnet/dotnet [0]
      2383:     find library=libgcc_s.so.1 [0]; searching
      2383:      search cache=/etc/ld.so.cache
      2383:       trying file=/lib/x86_64-linux-gnu/libgcc_s.so.1
      2383:
      2383:     file=libgcc_s.so.1 [0];  generating link map
      2383:       dynamic: 0x00007fa1ea07ede0  base: 0x00007fa1e9e68000   size: 0x0000000000217450
      2383:         entry: 0x00007fa1e9e6ab30  phdr: 0x00007fa1e9e68040  phnum:                  7
      2383:
      2383:
      2383:     file=libc.so.6 [0];  needed by /usr/share/dotnet/dotnet [0]
      2383:     find library=libc.so.6 [0]; searching
      2383:      search cache=/etc/ld.so.cache
      2383:       trying file=/lib/x86_64-linux-gnu/libc.so.6
      2383:
      2383:     file=libc.so.6 [0];  generating link map
      2383:       dynamic: 0x00007fa1e9e61b80  base: 0x00007fa1e9a77000   size: 0x00000000003f0ae0
      2383:         entry: 0x00007fa1e9a98cb0  phdr: 0x00007fa1e9a77040  phnum:                 10
      2383:
      2383:     checking for version `GCC_3.0' in file /lib/x86_64-linux-gnu/libgcc_s.so.1 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib/x86_64-linux-gnu/libdl.so.2 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.3' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib/x86_64-linux-gnu/libpthread.so.0 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.14' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.4' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.3' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBCXX_3.4.9' in file /usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `CXXABI_1.3' in file /usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBCXX_3.4.11' in file /usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBCXX_3.4' in file /usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0] required by file /usr/share/dotnet/dotnet [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_PRIVATE' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_2.14' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_2.3.2' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_2.4' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_PRIVATE' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2383:     checking for version `GLIBC_PRIVATE' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /lib/x86_64-linux-gnu/libdl.so.2 [0]
      2383:     checking for version `GLIBC_PRIVATE' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libdl.so.2 [0]
      2383:     checking for version `GLIBC_2.4' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libdl.so.2 [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib/x86_64-linux-gnu/libc.so.6 [0] required by file /lib/x86_64-linux-gnu/libdl.so.2 [0]
      2383:     checking for version `GLIBC_2.2.5' in file /lib/x86_64-linux-gnu/libm.so.6 [0] required by file /usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0]
      2383:     checking for version `GLIBC_2.3' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0]
      2383:     checking for version `GCC_4.2.0' in file
      2367:     symbol=ceil;  lookup in file=/usr/share/dotnet/shared/Microsoft.NETCore.App/3.1.0/libcoreclr.so [0]
      2367:     symbol=ceil;  lookup in file=dotnet [0]
      2367:     symbol=ceil;  lookup in file=/lib/x86_64-linux-gnu/libpthread.so.0 [0]
      2367:     symbol=ceil;  lookup in file=/lib/x86_64-linux-gnu/libdl.so.2 [0]
      2367:     symbol=ceil;  lookup in file=/usr/lib/x86_64-linux-gnu/libstdc++.so.6 [0]
      2367:     symbol=ceil;  lookup in file=/lib/x86_64-linux-gnu/libm.so.6 [0]
      2367:     binding file /usr/share/dotnet/shared/Microsoft.NETCore.App/3.1.0/libcoreclr.so [0] to /lib/x86_64-linux-gnu/libm.so.6 [0]: normal symbol `ceil' [GLIBC_2.2.5]

Test Run Aborted.
```
This system has glibc 2.27 on it:
```
01-09 22:48:34 10> ldd --version
ldd (Ubuntu GLIBC 2.27-3ubuntu1) 2.27
```
We did just switch from .NET Core SDK 2.2.203 to 3.1.100 and update the xUnit NuGet pkgs to:
```
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="coverlet.collector" Version="1.2.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
```

_Copied from original issue: dotnet/sdk#4150_

**Comment by ericstj** [comment]
@rkeithhill I'm not too familiar with these loader logs, but I didn't notice an error in that trace.  Can you point it out?  /cc @janvorli

**Comment by janvorli** [comment]
I don't see any error in the LD_DEBUG log either. To debug such a crash, it would be ideal to enable core dumps. Then when your test crashes, you'd be able to view the crash call stack using lldb and see both where it came from and what kind of crash it was. 
To do that, you need to make sure the core files are dumped for all apps (Ubuntu by default redirects failures to its own tool called apport and reports only issues in the apps). 
```sh
sysctl -w kernel.core_pattern=/tmp/core.%p
```
This command generates core files in the /tmp folder and names them core.PID where PID is the crashing process id. 
Maybe it would make sense to cat the previous value before modifying it so that you can restore it later if you want to.

Run these in the same console where you run the tests before starting your test:
```sh
ulimit -c unlimited
echo 0x3f > /proc/self/coredump_filter
```
The second command makes the core files to contain executable files data too, which makes debugging easier.

Now you can run your test and you should find a core file created in the /tmp folder. You can open it using lldb or gdb.

**Comment by rkeithhill** [comment]
I was able to analyze the apport generated crash dump. This is that I see:
```
Core was generated by `/usr/share/dotnet/dotnet exec --runtimeconfig /home/blduser/builds/foo/any-PR-U'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f351d6131a7 in std::vector<std::filesystem::__cxx11::path::_Cmpt, std::allocator<std::filesystem::__cxx11::path::_Cmpt> >::~vector() ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
[Current thread is 1 (Thread 0x7f35afb5e700 (LWP 86972))]
(gdb) where
#0  0x00007f351d6131a7 in std::vector<std::filesystem::__cxx11::path::_Cmpt, std::allocator<std::filesystem::__cxx11::path::_Cmpt> >::~vector() ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#1  0x00007f351d613231 in std::filesystem::__cxx11::path::~path() ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#2  0x00007f351d5f90b1 in Acme::FooClient::Internal::installFolderBarPath[abi:cxx11]() [clone .cold.97] () from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#3  0x00007f351d70f8fd in Acme::FooClient::Internal::bList1Path() ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#4  0x00007f351d70fc70 in Acme::FooClient::Internal::BCache::BCache() ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#5  0x00007f351d70fdde in Acme::FooClient::Internal::BCache::make() ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#6  0x00007f351d72e855 in Acme::FooClient::Internal::TransportableFlexConfigFilter::TransportableFlexConfigFilter(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Acme::FooClient::FooDesktopBarFormat) ()
   from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#7  0x00007f351d60eb02 in Acme::FooClient::FooDesktopClient::State::State(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Acme::FooClient::FooDesktopBarFormat)     () from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
#8  0x00007f351d60d03c in Acme::FooClient::FooDesktopClient::FooDesktopClient(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Acme::FooClient::FooDesktopBarFormat) () from /home/blduser/builds/foo/any-PR-Ubuntu/out/Linux-x64-Release/bin/FooApi.so
```

**Comment by janvorli** [comment]
So it is crashing in your native code.  I would recommend building the FooApi.so with debugging info enabled and without optimizations (`-g -O0`) and then running `/usr/share/dotnet/dotnet exec --runtimeconfig /home/blduser/builds/foo/any-PR-U` under gdb or lldb. The debugger would then break in right at the spot where the issue happened.
If you also need to see the managed frames on the stack, you'd need to use lldb, since gdb cannot walk such frames and stops at the last native frame.

**Comment by ericstj** [comment]
Closing as this was tracking an crash occuring in customer native code.  Please open a new issue if investigation determines otherwise.

Thanks!
